File: Linc/DESKTOP/Linc.Desktop/Protocol/Framing.cs
using System.Buffers.Binary;
using System.IO;
using System.Text;

namespace Linc.Desktop.Protocol;

/// <summary>
/// Frame codec: 4-byte unsigned big-endian length prefix + UTF-8 JSON (docs/PROTOCOL.md).
/// </summary>
public static class Framing
{
    public const int MaxFrameBytes = 1024 * 1024;

    public static async Task WriteAsync(Stream stream, string json, CancellationToken ct)
    {
        var bytes = Encoding.UTF8.GetBytes(json);
        if (bytes.Length > MaxFrameBytes)
        {
            throw new IOException($"outgoing frame of {bytes.Length} bytes exceeds {MaxFrameBytes}");
        }
        var prefix = new byte[4];
        BinaryPrimitives.WriteInt32BigEndian(prefix, bytes.Length);
        await stream.WriteAsync(prefix, ct);
        await stream.WriteAsync(bytes, ct);
        await stream.FlushAsync(ct);
    }

    /// <summary>Reads one frame, or returns null on clean end-of-stream.</summary>
    public static async Task<string?> ReadAsync(Stream stream, CancellationToken ct)
    {
        var prefix = new byte[4];
        var first = await stream.ReadAsync(prefix.AsMemory(0, 4), ct);
        if (first == 0)
        {
            return null;
        }
        if (first < 4)
        {
            await stream.ReadExactlyAsync(prefix.AsMemory(first, 4 - first), ct);
        }
        var length = BinaryPrimitives.ReadInt32BigEndian(prefix);
        if (length <= 0 || length > MaxFrameBytes)
        {
            throw new IOException($"invalid frame length: {length}");
        }
        var buffer = new byte[length];
        await stream.ReadExactlyAsync(buffer, ct);
        return Encoding.UTF8.GetString(buffer);
    }

    /// <summary>
    /// Reads one raw binary frame (bulk channel, docs/PROTOCOL.md v5), or returns null on
    /// clean end-of-stream. Unlike <see cref="ReadAsync"/>, a zero-length frame is legal
    /// here and returns an empty array ("not found / unavailable").
    /// </summary>
    public static async Task<byte[]?> ReadBytesAsync(Stream stream, CancellationToken ct)
    {
        var prefix = new byte[4];
        var first = await stream.ReadAsync(prefix.AsMemory(0, 4), ct);
        if (first == 0)
        {
            return null;
        }
        if (first < 4)
        {
            await stream.ReadExactlyAsync(prefix.AsMemory(first, 4 - first), ct);
        }
        var length = BinaryPrimitives.ReadInt32BigEndian(prefix);
        if (length < 0 || length > MaxFrameBytes)
        {
            throw new IOException($"invalid frame length: {length}");
        }
        if (length == 0)
        {
            return [];
        }
        var buffer = new byte[length];
        await stream.ReadExactlyAsync(buffer, ct);
        return buffer;
    }
}

File: Linc/ANDROID/app/src/main/java/app/linc/android/protocol/Framing.kt
package app.linc.android.protocol

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException
import java.io.IOException

/**
 * Frame codec: 4-byte unsigned big-endian length prefix + UTF-8 JSON (docs/PROTOCOL.md).
 * DataOutputStream.writeInt is big-endian, matching the spec; lengths are capped at
 * 1 MiB, so the sign bit is never set and signed readInt is safe.
 */
object Framing {
    const val MAX_FRAME_BYTES = 1 * 1024 * 1024

    fun write(out: DataOutputStream, json: String) {
        val bytes = json.toByteArray(Charsets.UTF_8)
        if (bytes.size > MAX_FRAME_BYTES) {
            throw IOException("outgoing frame of ${bytes.size} bytes exceeds $MAX_FRAME_BYTES")
        }
        out.writeInt(bytes.size)
        out.write(bytes)
        out.flush()
    }

    /** Reads one frame, or returns null on clean end-of-stream. */
    fun read(input: DataInputStream): String? {
        val length = try {
            input.readInt()
        } catch (_: EOFException) {
            return null
        }
        if (length <= 0 || length > MAX_FRAME_BYTES) {
            throw IOException("invalid frame length: $length")
        }
        val buffer = ByteArray(length)
        input.readFully(buffer)
        return String(buffer, Charsets.UTF_8)
    }

    /**
     * Writes one raw binary frame (bulk channel, PROTOCOL.md v5). Same 4-byte
     * big-endian length prefix as [write]; a zero-length frame is legal and means
     * "not found / unavailable".
     */
    fun writeBytes(out: DataOutputStream, bytes: ByteArray) {
        if (bytes.size > MAX_FRAME_BYTES) {
            throw IOException("outgoing frame of ${bytes.size} bytes exceeds $MAX_FRAME_BYTES")
        }
        out.writeInt(bytes.size)
        out.write(bytes)
        out.flush()
    }

    /** Reads one raw binary frame, or returns null on clean end-of-stream. */
    fun readBytes(input: DataInputStream): ByteArray? {
        val length = try {
            input.readInt()
        } catch (_: EOFException) {
            return null
        }
        if (length < 0 || length > MAX_FRAME_BYTES) {
            throw IOException("invalid frame length: $length")
        }
        val buffer = ByteArray(length)
        input.readFully(buffer)
        return buffer
    }
}
